File: binary-search-tree/BinarySearchTree.kt
class Node(var data: Int) {
    var left: Node? = null
    var right: Node? = null
}

class BST {
    private var root: Node? = null
    
    fun insert(data: Int) {
        root = insertRec(root, data)
    }
    
    private fun insertRec(root: Node?, data: Int): Node {
        if (root == null) return Node(data)
        
        if (data < root.data) root.left = insertRec(root.left, data)
        else root.right = insertRec(root.right, data)
        
        return root
    }
    
    fun search(data: Int): Boolean = searchRec(root, data)
    
    private fun searchRec(root: Node?, data: Int): Boolean {
        if (root == null) return false
        if (root.data == data) return true
        return if (data < root.data) searchRec(root.left, data) else searchRec(root.right, data)
    }
}

fun main() {
    val bst = BST()
    listOf(50, 30, 70, 20, 40).forEach { bst.insert(it) }
    println(bst.search(40))
}

File: binary-search-tree/binary_search_tree.ts
class TreeNode {
    data: number;
    left: TreeNode | null = null;
    right: TreeNode | null = null;
    
    constructor(data: number) {
        this.data = data;
    }
}

class BST {
    private root: TreeNode | null = null;
    
    insert(data: number): void {
        this.root = this.insertRec(this.root, data);
    }
    
    private insertRec(node: TreeNode | null, data: number): TreeNode {
        if (!node) return new TreeNode(data);
        
        if (data < node.data) node.left = this.insertRec(node.left, data);
        else node.right = this.insertRec(node.right, data);
        
        return node;
    }
    
    search(data: number): boolean {
        return this.searchRec(this.root, data);
    }
    
    private searchRec(node: TreeNode | null, data: number): boolean {
        if (!node) return false;
        if (node.data === data) return true;
        return data < node.data ? this.searchRec(node.left, data) : this.searchRec(node.right, data);
    }
}

const bst = new BST();
[50, 30, 70, 20, 40].forEach(x => bst.insert(x));
console.log(bst.search(40));